### ISSUE TITLE
Running Euroscope via OpenFSD

### ISSUE THREAD

**Issue by trinhxuanson** [renorris/openfsd#6]
I tried to run the OpenFSD successully; I wonder if if I can run Euroscope via this FSD server? since the built-in FSD server of Euroscope only restricted 2 controller connections only.

**Comment by renorris** [comment]
Currently, ATC connections aren't supported. Eventually it may be added; I'll keep this issue open as a feature request.

**Comment by JianyueLab** [comment]
When will ATC connect be available

**Comment by fwilliamhe** [comment]
> Currently, ATC connections aren't supported. Eventually it may be added; I'll keep this issue open as a feature request.

There is a way to work with Euroscope by replacing the `URL` in Euroscope, but it may be against EULA.

**Comment by renorris** [comment]
> > Currently, ATC connections aren't supported. Eventually it may be added; I'll keep this issue open as a feature request.
> 
> 
> 
> There is a way to work with Euroscope by replacing the `URL` in Euroscope, but it may be against EULA.

Maybe so, but to be clear: there's no openfsd support for the ATC-specific FSD protocol. There are also other API endpoints hardcoded into Euroscope & other clients (aside from the plain FSD server address) that need to be addressed as well.

**Comment by fwilliamhe** [comment]
> > > Currently, ATC connections aren't supported. Eventually it may be added; I'll keep this issue open as a feature request.
> > 
> > 
> > There is a way to work with Euroscope by replacing the `URL` in Euroscope, but it may be against EULA.
> 
> Maybe so, but to be clear: there's no openfsd support for the ATC-specific FSD protocol. There are also other API endpoints hardcoded into Euroscope & other clients (aside from the plain FSD server address) that need to be addressed as well.

Yeah. But I have implemented some protocols  with C++, I can share them with you.

**Comment by trinhxuanson** [comment]
> > > > Currently, ATC connections aren't supported. Eventually it may be added; I'll keep this issue open as a feature request.
> > > 
> > > 
> > > There is a way to work with Euroscope by replacing the `URL` in Euroscope, but it may be against EULA.
> > 
> > 
> > Maybe so, but to be clear: there's no openfsd support for the ATC-specific FSD protocol. There are also other API endpoints hardcoded into Euroscope & other clients (aside from the plain FSD server address) that need to be addressed as well.
> 
> Yeah. But I have implemented some protocols with C++, I can share them with you.

Yeahhh, could you share for me please l, i really want to test the FSD server for ATC clients

**Comment by renorris** [comment]
@Misaka-Nnnnq Feel free to email whatever you'd like to mail (at) reesenorr.is.

Full protocol documentation is in the works on my end. Will be shared once complete & used for the next openfsd iteration.

**Comment by JianyueLab** [comment]
> > > > Currently, ATC connections aren't supported. Eventually it may be added; I'll keep this issue open as a feature request.
> > > 
> > > 
> > > There is a way to work with Euroscope by replacing the `URL` in Euroscope, but it may be against EULA.
> > 
> > 
> > Maybe so, but to be clear: there's no openfsd support for the ATC-specific FSD protocol. There are also other API endpoints hardcoded into Euroscope & other clients (aside from the plain FSD server address) that need to be addressed as well.
> 
> Yeah. But I have implemented some protocols with C++, I can share them with you.

May I have one copy plz.

**Comment by michael-filhol** [comment]
Hello,
I was going to contact the author to know more about this project's user base, usage, support, etc. because:
  1. I thought plain old FSD (i.e. outside of Vatsim) was barely living any more, and virtually only hanging on the old kuroneko/fsd fork, whereas this repo looks pretty active and relevant;
  2. I authored and maintain an ATC simulation program (ATC-pie) which can connect to FSD networks, at least those served by the kurenoko version, a feature I added as I met a flight sim group that took interest in it.

So this begs questions from me... Is openfsd meant to be a continuation of FSD, or does it redefine things fresh or in a non-compatible way?

In the former case, ATC-pie could be the already available solution that people in this thread are looking for (replacing Euroscope). In the latter, I would be very interested in studying renorris's work here and why not add yet another connection type in ATC-pie to support ATC roles on openfsd networks.

**Comment by renorris** [comment]
1. Yes, FSD is barely living anymore outside of VATSIM and IVAO. (Not even sure about IVAO... do they still use FSD?)
This is a personal project that I work on during my free time (which is quite limited these days.) I just think its cool, other than that there's no real reason.
Since the last major public release, there's been lots of programming behind the scenes working towards a fully-functional ATC system. No ETA at all for this though.

2.  Very cool. I will check out your ATC client and see how I can support it one day. Is there any way to view it?

The philosophy of openfsd is simply meant to be an open-source alternative *reimplementing* VATSIM's version of FSD. There's no major reinventing the wheel or "redefining" anything. If a VATSIM client exists, I'd eventually like to support it on openfsd. I don't intend to "replace" Euroscope, or any other client. I want to support them.

**Comment by michael-filhol** [comment]
Of course you can have a look: see the [project page](https://sourceforge.net/projects/atc-pie), and a description of the features on the [wiki page](https://wiki.flightgear.org/ATC-pie). 
My point was that ATC-pie might already work transparently, regardless of Euroscope, provided that you have as little as position packet forwarding and the text chat system covered. This even includes the top features such as 3D tower view (through FlightGear), CPDLC (through Hoppie like Vatsim does), and phone land lines. The last section of the wiki page is a note on interoperability with Euroscope. It is limited, but programs can still coexist. I am happy to extend support here if you provide a way to work around their NDA, but I don't think this is possible.
We can take this to a separate thread as soon as you believe it becomes appropriate (e.g. off topic here).

**Comment by renorris** [comment]
> Of course you can have a look: see the [project page](https://sourceforge.net/projects/atc-pie), and a description of the features on the [wiki page](https://wiki.flightgear.org/ATC-pie). My point was that ATC-pie might already work transparently, regardless of Euroscope, provided that you have as little as position packet forwarding and the text chat system covered. This even includes the top features such as 3D tower view (through FlightGear), CPDLC (through Hoppie like Vatsim does), and phone land lines. The last section of the wiki page is a note on interoperability with Euroscope. It is limited, but programs can still coexist. I am happy to extend support here if you provide a way to work around their NDA, but I don't think this is possible. We can take this to a separate thread as soon as you believe it becomes appropriate (e.g. off topic here).

@michael-filhol Please see the new FSD documentation found [here](https://github.com/renorris/openfsd/tree/main/docs). You can view the .md files directly or build the documentation via MkDocs using the `mkdocs.yml` found in the root directory of the repository. This should include all you need to add openfsd support to ATC-pie. Let me know if you have any questions!

**Comment by renorris** [comment]
> > Euroscope is supported by v1.0-beta. The client itself requires a patch directly to the binary, and the tooling for this is currently WIP. This thread will be updated.
> 
> Sorry for my late response. I'm busy with my study. I can upload an example of the patch for you.

Sure, why not.

**Comment by Misaka-Nnnnq** [comment]
> > > Euroscope is supported by v1.0-beta. The client itself requires a patch directly to the binary, and the tooling for this is currently WIP. This thread will be updated.
> > 
> > 
> > Sorry for my late response. I'm busy with my study. I can upload an example of the patch for you.
> 
> Sure, why not.

I have created a for [repo](https://github.com/Misaka-Nnnnq/openfsd-patch-for-es) for that.
A config tool and new version is currently in development and I'm very glad to make it for openfsd.

**Comment by renorris** [comment]
Excellent work—I also have some work done on a binary patcher (not runtime) where a longer URL can be used as well. Once finished I'll share it. I already have the offsets that will need to be changed:

```
===== Euroscope 3.2.9 =====

SHA1	dfb1caf3d73e897b2a04964dc35867b4059bc537

Raw offsets:
Segment	Raw Offset	Virtual Start
.text	0x400		0x401000
.data	0x2E5000	0x6E7000

Disable FSD version check:
.text	0x5AE01A	Replace 0x74 with 0xEB

Write fsd-jwt URL into unused db slot:
Unused string slot:	.rdata:0x65DE58		Length: 0x76

Change fsd-jwt push offset:
.text	0x4644E3	Overwrite 3 bytes: 58 DE 65
```

**Comment by renorris** [comment]
@Misaka-Nnnnq See [here](https://github.com/renorris/openfsd-client-patch-utility) for some preliminary work I've done with Euroscope.

**Comment by renorris** [comment]
It's in the wiki. If you've already followed what's in the wiki and it still doesn't work, let me know.

https://github.com/renorris/openfsd-client-patch-utility
https://github.com/renorris/openfsd-client-patch-utility/wiki/Euroscope

**Comment by trinhxuanson** [comment]
<img width="618" height="861" alt="Image" src="https://github.com/user-attachments/assets/ee7094d7-f549-4320-96f1-67fd384525de" />

<img width="734" height="849" alt="Image" src="https://github.com/user-attachments/assets/37133625-b17b-4eef-8daa-d67ae550be78" />

<img width="941" height="399" alt="Image" src="https://github.com/user-attachments/assets/c490e8b9-fbb9-4ff2-a15c-6c4d683bd11f" />

<img width="415" height="64" alt="Image" src="https://github.com/user-attachments/assets/921c16b0-4844-4221-8cd1-3184181d05b8" />


I tried to follow the openfsd client patch utility already but same problems remain

**Comment by Olivender** [comment]
Try to make without the api on your domain like https://xsfsd.com/ ....

**Comment by trinhxuanson** [comment]
could you show me the step-by-step tutorial?
I'm a newbie. I'm trying to run the FSD for Euroscope but nothing works. 
Honestly; I have no idea how to run this even though reading through the wiki
